src/pytorch_lightning/strategies/colossalai.py

```diff
@@ -0,0 +1,275 @@
+from typing import Any, Generator, Optional
+
+import torch
+from lightning_utilities.core.imports import RequirementCache
+
+import pytorch_lightning as pl
+from pytorch_lightning.accelerators.cuda import CUDAAccelerator
+from pytorch_lightning.overrides.base import _LightningModuleWrapperBase, _LightningPrecisionModuleWrapperBase
+from pytorch_lightning.plugins.precision import ColossalAIPrecisionPlugin
+from pytorch_lightning.strategies.ddp import DDPStrategy
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.types import STEP_OUTPUT
+
+_COLOSSALAI_AVAILABLE = RequirementCache("colossalai")
+if _COLOSSALAI_AVAILABLE:
+    from colossalai.context import ParallelMode
+    from colossalai.core import global_context as gpc
+    from colossalai.gemini import ChunkManager, GeminiManager
+    from colossalai.logging import disable_existing_loggers, get_dist_logger
+    from colossalai.nn.optimizer import CPUAdam, HybridAdam
+    from colossalai.nn.parallel import ZeroDDP
+    from colossalai.tensor import ProcessGroup
+    from colossalai.utils import get_current_device
+    from colossalai.utils.model.colo_init_context import ColoInitContext
+    from colossalai.zero import ZeroOptimizer
+
+
+class ModelShardedContext(ColoInitContext):
+    def _post_init_method(self, module: torch.nn.Module, *args, **kwargs):
+        if getattr(module, "_colossalai_module", False) is True:
+            return
+        super()._post_init_method(module, *args, **kwargs)
+        module._colossalai_module = True
+
+
+class ColossalAIStrategy(DDPStrategy):
+    """ColossalAI strategy. It only supports single optimizer which must be  `colossalai.nn.optimizer.CPUAdam`_ or
+    `colossalai.nn.optimizer.HybridAdam`_ now. You must initialize your model in ``configure_sharded_model()``.
+
+    It configures accelerator and precision, and you should not configure them when initializing ``Trainer``.
+    CUDA is essential for this strategy. Please make sure CUDA is available.
+
+    Example::
+
+        class GLUETransformer(LightningModule):
+            ...
+            def configure_sharded_model(self) -> None:
+                self.model = BertForSequenceClassification.from_pretrained('bert-base-uncased')
```

Comment:
this is an example of manual sharding. Does it support auto-sharding?

Reply:
It is not an example of manual sharding. The fact is that we require users to initialize their models in the context `ColoInitContext`, which is wrapped as the return value of `model_sharded_context` function in the strategy. As a result, we set the model attribute of `GLUETransformer` as None in its `__init__` function, and use `configure_sharded_model` function where the model is actually initialized. 
Although Colossalai strategy always make the model auto-sharded, users do not need to care about the distributed storage of the model. We hide all sharding details away from users. 
In sum, we wrote like this to force users initializing their models in `ColoInitContext`. Do you have any idea to realize it in a more explicit way?

Reply:
if users send a LightningModule initialized on CPU containing some layers inside already and not overriding the `configure_sharded_model`, is it possible for colossalai to initialize it somehow without any context, with some sort of wrapper? Something we recently added to [fully sharded native recently](https://github.com/Lightning-AI/lightning/pull/14252).

Reply:
Currently, we don't support transforming models by some wrappers. It is necessary for users to initialize models with the context `ColoInitContext`.